zerver/actions/uploads.py

```diff
@@ -85,40 +88,56 @@ def do_delete_old_unclaimed_attachments(weeks_ago: int) -> None:
         weeks_ago
     )
 
+    old_unclaimed_attachments.update(deleted=True)
+    old_unclaimed_archived_attachments.update(deleted=True)
+
+    attachments_path_ids = [attachment.path_id for attachment in old_unclaimed_attachments]
+
+    archived_attachments_path_ids = [
+        archived_attachment.path_id for archived_attachment in old_unclaimed_archived_attachments
+    ]
+
+    ImageAttachment.objects.filter(
+        Q(path_id__in=attachments_path_ids) | Q(path_id__in=archived_attachments_path_ids)
+    ).update(deleted=True)
+
     # An attachment may be removed from Attachments and
     # ArchiveAttachments in the same run; prevent warnings from the
     # backing store by only removing it from there once.
     already_removed = set()
     storage_paths = []
-    for attachment in old_unclaimed_attachments:
-        storage_paths.append(attachment.path_id)
-        image_row = ImageAttachment.objects.filter(path_id=attachment.path_id).first()
+    for attachment_path_id in attachments_path_ids:
+        storage_paths.append(attachment_path_id)
+        image_row = ImageAttachment.objects.filter(path_id=attachment_path_id).first()
         if image_row:
             for existing_thumbnail in image_row.thumbnail_metadata:
                 thumb = StoredThumbnailFormat(**existing_thumbnail)
                 storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-            image_row.delete()
-        already_removed.add(attachment.path_id)
-        attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-    for archived_attachment in old_unclaimed_archived_attachments:
-        if archived_attachment.path_id not in already_removed:
-            storage_paths.append(archived_attachment.path_id)
-            image_row = ImageAttachment.objects.filter(path_id=archived_attachment.path_id).first()
+        already_removed.add(attachment_path_id)
+
+    for archived_attachment_path_id in archived_attachments_path_ids:
+        if archived_attachment_path_id not in already_removed:
+            storage_paths.append(archived_attachment_path_id)
+            image_row = ImageAttachment.objects.filter(path_id=archived_attachment_path_id).first()
             if image_row:  # nocoverage
                 for existing_thumbnail in image_row.thumbnail_metadata:
                     thumb = StoredThumbnailFormat(**existing_thumbnail)
                     storage_paths.append(get_image_thumbnail_path(image_row, thumb))
-                image_row.delete()
-        archived_attachment.delete()
-        if len(storage_paths) >= DELETE_BATCH_SIZE:
-            delete_message_attachments(storage_paths[:DELETE_BATCH_SIZE])
-            storage_paths = storage_paths[DELETE_BATCH_SIZE:]
-
-    if storage_paths:
-        delete_message_attachments(storage_paths)
+
```

Comment:
The logic above likely should be querying everything with `deleted=True`, not the set of things passed in, if we want failures to be handled properly, right?

Reply:
You mean instead of passing `storage_paths` we should pass `path_id`s of all `deleted=True` objects? 

But there are some `path_id`s that don't have a corresponding db object, like [thumbnail path](https://github.com/zulip/zulip/blob/main/zerver/actions/uploads.py#L98-L99)